tests/io/test_data_catalog.py

```diff
@@ -683,3 +745,109 @@ def test_no_versions_with_cloud_protocol(self):
         )
         with pytest.raises(DataSetError, match=pattern):
             versioned_dataset.load()
+
+
+class TestDataCatalogDatasetFactories:
+    def test_match_added_to_datasets_on_get(self, config_with_dataset_factories):
+        """Check that the datasets that match patterns are only added when fetched"""
+        catalog = DataCatalog.from_config(**config_with_dataset_factories)
+        assert "{brand}_cars" not in catalog._data_sets
+        assert "tesla_cars" not in catalog._data_sets
+        assert "{brand}_cars" in catalog.dataset_patterns
+
+        tesla_cars = catalog._get_dataset("tesla_cars")
+        assert isinstance(tesla_cars, CSVDataSet)
+        assert "tesla_cars" in catalog._data_sets
+
+    @pytest.mark.parametrize(
+        "dataset_name, expected",
+        [
+            ("audi_cars", True),
+            ("tesla_cars", True),
+            ("row_boats", True),
+            ("boats", False),
+            ("tesla_card", False),
+        ],
+    )
+    def test_exists_in_catalog(
+        self, config_with_dataset_factories, dataset_name, expected
+    ):
+        """Check that the dataset exists in catalog when it matches a pattern
+        or is in the catalog"""
+        catalog = DataCatalog.from_config(**config_with_dataset_factories)
+        assert catalog.exists_in_catalog(dataset_name) == expected
+
+    @pytest.mark.parametrize(
+        "dataset_name, expected",
+        [
+            ("tesla_cars", True),
+            ("row_boats", True),
+            ("boats", False),
+            ("ford_cars", True),
+            ("sail_boats", True),
+        ],
+    )
+    def test_exists_in_catalog_cache_updated(
+        self, config_with_dataset_factories, dataset_name, expected
+    ):
+        """Check that the pattern match cache is updated when exists_in_catalog() is called"""
+        catalog = DataCatalog.from_config(**config_with_dataset_factories)
+        assert catalog.exists_in_catalog(dataset_name) == expected
```

Comment:
I don't quite understand the difference between this test and the previous one.

Reply:
The one before this was meant to check if the fn returns true or false correctly for datasets that exist either as explicit entries or patterns. This one is meant to check if the cache is updated correctly so all the examples here are just pattern matches. Updated it!